homie-influx/src/main.rs

```diff
@@ -0,0 +1,388 @@
+    device_id: String,
+    node_id: String,
+    property_id: String,
+) {
+    if let Some(device) = controller.devices().get(&device_id) {
+        if let Some(node) = device.nodes.get(&node_id) {
+            if let Some(property) = node.properties.get(&property_id) {
+                if let Some(point) =
+                    point_for_property_value(device, node, property, SystemTime::now())
+                {
+                    // TODO: What should rp be?
+                    // TODO: Handle errors
+                    influx_db_client
+                        .write_point(point, INFLUXDB_PRECISION, None)
+                        .await
+                        .unwrap();
+                }
+            }
+        }
+    }
+}
+
+/// Convert the value of the given Homie property to an InfluxDB value of the appropriate type, if
+/// possible. Returns None if the datatype of the property is unknown, or there was an error parsing
+/// the value.
+fn influx_value_for_homie_property(property: &Property) -> Option<Value> {
+    let datatype = property.datatype?;
+    Some(match datatype {
+        Datatype::Integer => Value::Integer(property.value().ok()?),
+        Datatype::Float => Value::Float(property.value().ok()?),
+        Datatype::Boolean => Value::Boolean(property.value().ok()?),
+        _ => Value::String(property.value.to_owned()?),
+    })
+}
+
+/// Construct an InfluxDB `Point` corresponding to the given Homie property value update.
+fn point_for_property_value(
+    device: &Device,
+    node: &Node,
+    property: &Property,
+    timestamp: SystemTime,
+) -> Option<Point> {
+    let datatype = property.datatype?;
+    let value = influx_value_for_homie_property(property)?;
+
+    let mut point = Point::new(&datatype.to_string())
+        .add_timestamp(
+            timestamp
+                .duration_since(SystemTime::UNIX_EPOCH)
+                .unwrap()
+                .as_millis() as i64,
+        )
+        .add_field("value", value)
+        .add_tag("device_id", Value::String(device.id.to_owned()))
+        .add_tag("node_id", Value::String(node.id.to_owned()))
+        .add_tag("property_id", Value::String(property.id.to_owned()));
+    if let Some(device_name) = device.name.to_owned() {
+        point = point.add_field("device_name", Value::String(device_name));
+    }
+    if let Some(node_name) = node.name.to_owned() {
+        point = point.add_field("node_name", Value::String(node_name));
+    }
+    if let Some(property_name) = property.name.to_owned() {
+        point = point.add_field("property_name", Value::String(property_name));
+    }
+    if let Some(unit) = property.unit.to_owned() {
+        point = point.add_tag("unit", Value::String(unit));
+    }
+    if let Some(node_type) = node.node_type.to_owned() {
+        point = point.add_tag("node_type", Value::String(node_type))
+    }
+
+    Some(point)
+}
+
+fn simplify_unit_vec<E>(m: Result<Vec<()>, E>) -> Result<(), E> {
+    m.map(|_| ())
+}
+
```

Comment:
I know that the other binary's main.rs is just as big, but it's shocking seeing all of this code appear in one file in one PR.

Can we split out a `config.rs` and an `influx.rs` so that main.rs only contains `main()` and maybe `spawn_homie_poll_loop()` and `simplify_unit_vec()`?

I can do the mijia-homie refactor as a separate PR if you want.

Reply:
Done.